lib/Dialect/Calyx/Transforms/AffineParallelUnroll.cpp

```diff
@@ -45,201 +83,151 @@ namespace {
-  for (auto &[memOp, constIndices] : constantMemAccessIndices) {
-    auto readOp = dyn_cast<AffineReadOpInterface>(memOp);
-    if (!readOp)
-      continue;
+bool MemoryBankConflictResolver::isInvariantToAffineParallel(
+    AffineReadOpInterface readOp, AffineParallelOp affineParallelOp) {
+  // Check if any operand depends on loop IVs
+  for (Value iv : affineParallelOp.getIVs()) {
+    for (Value operand : readOp.getMapOperands()) {
+      if (operand == iv)
+        return false;
+
+      // Walk through defining operation chains, such as `affine.apply`s.
+      if (Operation *def = operand.getDefiningOp()) {
+        if (affineParallelOp->isAncestor(def) && !isa<arith::ConstantOp>(def)) {
+          // Operand is computed inside the loop, not invariant
+          return false;
+        }
+      }
+    }
+  }
 
-    auto memref = readOp.getMemRef();
-    auto key = std::pair(memref, constIndices);
-    // We do not hoist any read as long as it's being written in any parallel
-    // region.
-    if (llvm::any_of(memref.getUsers(), [&](Operation *user) {
-          return affineParallelOp->isAncestor(user) &&
-                 hasEffect<MemoryEffects::Write>(user, memref);
-        })) {
+  // Check if memref is written in loop
+  Value memref = readOp.getMemRef();
+  for (Operation *user : memref.getUsers()) {
+    if (user == readOp.getOperation())
       continue;
-    }
-    constantReadOpCounts[key]++;
+
+    if (affineParallelOp->isAncestor(user) &&
+        hasEffect<MemoryEffects::Write>(user, memref))
+      return false;
   }
 
-  bool shouldHoist = llvm::any_of(
-      constantReadOpCounts, [](const auto &entry) { return entry.second > 1; });
-  if (!shouldHoist)
-    return;
+  return true;
+}
 
+LogicalResult
+MemoryBankConflictResolver::readOpAnalysis(AffineParallelOp affineParallelOp) {
   OpBuilder builder(affineParallelOp);
-  DenseMap<std::pair<Value, SmallVector<int64_t, 4>>, ValueRange> hoistedReads;
-  for (auto &[memOp, constIndices] : constantMemAccessIndices) {
-    auto readOp = dyn_cast<AffineReadOpInterface>(memOp);
-    if (!readOp)
+  for (auto readOp : allReadOps) {
+    AffineAccessExpr key{readOp.getMemRef(), readOp.getAffineMap(),
+                         readOp.getMapOperands()};
+    auto it = readOpCounts.find(key);
+    if (it == readOpCounts.end() || it->second <= 1 ||
+        !isInvariantToAffineParallel(readOp, affineParallelOp))
       continue;
 
-    auto key = std::pair(readOp.getMemRef(), constIndices);
-    if (constantReadOpCounts[key] > 1) {
-      if (hoistedReads.find(key) == hoistedReads.end()) {
-        builder.setInsertionPoint(affineParallelOp);
-        Operation *clonedRead = builder.clone(*readOp.getOperation());
-        hoistedReads[key] = clonedRead->getOpResults();
-      }
-      readOp->replaceAllUsesWith(hoistedReads[key]);
+    // Only hoist once per unique access
+    if (!hoistedReads.contains(key)) {
```

Comment:
Can we use `.find(key)` please to avoid triple lookup (`contains`, then `[key]` twice below)?